components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/RoleBasedScopeIssuer.java

```diff
@@ -0,0 +1,758 @@
+        CacheEntry result = OAuthCache.getInstance().getValueFromCache(cacheKey);
+        // Cache hit
+        if (result instanceof ResourceScopeCacheEntry) {
+            resourceScope = ((ResourceScopeCacheEntry) result).getScope();
+            resourceTenantId = ((ResourceScopeCacheEntry) result).getTenantId();
+            cacheHit = true;
+        }
+        // Cache was not hit. So retrieve from database.
+        if (!cacheHit) {
+            Pair<String, Integer> scopeMap = OAuthTokenPersistenceFactory.getInstance()
+                    .getTokenManagementDAO().findTenantAndScopeOfResource(resource);
+            if (scopeMap != null) {
+                resourceScope = scopeMap.getLeft();
+                resourceTenantId = scopeMap.getRight();
+            }
+            cacheKey = new OAuthCacheKey(resource);
+            ResourceScopeCacheEntry cacheEntry = new ResourceScopeCacheEntry(resourceScope);
+            cacheEntry.setTenantId(resourceTenantId);
+            // Store resourceScope in cache even if it is null (to avoid database calls when accessing resources for
+            // which scopes haven't been defined).
+            OAuthCache.getInstance().addToCache(cacheKey, cacheEntry);
+        }
+        // Return TRUE if - There does not exist a scope definition for the resource.
+        if (resourceScope == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Resource '" + resource + "' is not protected with a scope");
+            }
+            return true;
+        }
+        List<String> scopeList = new ArrayList<>(Arrays.asList(scopes));
+        // If the access token does not bear the scope required for accessing the Resource.
+        if (!scopeList.contains(resourceScope)) {
+            if (log.isDebugEnabled() && IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.ACCESS_TOKEN)) {
+                log.debug("Access token '" + accessTokenDO.getAccessToken() + "' does not bear the scope '"
+                        + resourceScope + "'");
+            }
+            return false;
+        }
+        // If a federated user and CHECK_ROLES_FROM_SAML_ASSERTION system property is set to true,
+        // or if a federated user and RETRIEVE_ROLES_FROM_USERSTORE_FOR_SCOPE_VALIDATION system property is false,
+        // avoid validating user roles.
+        // This system property is set at server start using -D option, Thus will be a permanent property.
+        if (accessTokenDO.getAuthzUser().isFederatedUser()
+                && (Boolean.parseBoolean(System.getProperty(CHECK_ROLES_FROM_SAML_ASSERTION)) ||
+                !(Boolean.parseBoolean(System.getProperty(RETRIEVE_ROLES_FROM_USERSTORE_FOR_SCOPE_VALIDATION))))) {
+            return true;
+        }
+        AuthenticatedUser authenticatedUser = OAuthUtil
+                .getAuthenticatedUser(oAuth2TokenValidationMessageContext.getResponseDTO().getAuthorizedUser());
```

Comment:
can the `oAuth2TokenValidationMessageContext` be null?

Reply:
perhaps we could check that at the beginning?

if (oAuth2TokenValidationMessageContext == null) {
        return false;
    }

Reply:
It cannot be null, it is initalized properly in the calling class eg: TokenValidationHandler. IMO it would be wrong to do a null as it is a variable which should be initialized at the first place.